File: Python/03-test-generation/calculator.py
"""
Demo 3 — Test Generation
=========================
This module has fully-implemented business logic but ZERO tests.
Use Copilot to generate a comprehensive test suite.

Exercises:
  1. Open this file → Copilot Chat → "/tests" or "Generate unit tests for this file"
  2. Select a single method → Inline Chat → "Write pytest tests including edge cases"
  3. Ask Chat: "What edge cases am I missing?"
  4. Run the generated tests: pytest test_calculator.py -v
"""

import math


class Calculator:
    """A calculator with memory, history, and scientific operations."""

    def __init__(self):
        self.memory: float = 0.0
        self.history: list[str] = []

    def add(self, a: float, b: float) -> float:
        result = a + b
        self._record(f"{a} + {b} = {result}")
        return result

    def subtract(self, a: float, b: float) -> float:
        result = a - b
        self._record(f"{a} - {b} = {result}")
        return result

    def multiply(self, a: float, b: float) -> float:
        result = a * b
        self._record(f"{a} * {b} = {result}")
        return result

    def divide(self, a: float, b: float) -> float:
        if b == 0:
            raise ZeroDivisionError("Cannot divide by zero")
        result = a / b
        self._record(f"{a} / {b} = {result}")
        return result

    def power(self, base: float, exponent: float) -> float:
        result = math.pow(base, exponent)
        self._record(f"{base} ^ {exponent} = {result}")
        return result

    def sqrt(self, value: float) -> float:
        if value < 0:
            raise ValueError("Cannot compute square root of a negative number")
        result = math.sqrt(value)
        self._record(f"√{value} = {result}")
        return result

    def percentage(self, value: float, percent: float) -> float:
        """Return `percent`% of `value`. E.g. percentage(200, 15) → 30.0"""
        result = value * (percent / 100)
        self._record(f"{percent}% of {value} = {result}")
        return result

    def store(self, value: float) -> None:
        """Save a value to memory."""
        self.memory = value

    def recall(self) -> float:
        """Retrieve the stored memory value."""
        return self.memory

    def clear_memory(self) -> None:
        self.memory = 0.0

    def get_history(self) -> list[str]:
        return list(self.history)

    def clear_history(self) -> None:
        self.history.clear()

    def _record(self, entry: str) -> None:
        self.history.append(entry)


class UnitConverter:
    """Converts between common units."""

    @staticmethod
    def celsius_to_fahrenheit(c: float) -> float:
        return (c * 9 / 5) + 32

    @staticmethod
    def fahrenheit_to_celsius(f: float) -> float:
        return (f - 32) * 5 / 9

    @staticmethod
    def kg_to_lbs(kg: float) -> float:
        return kg * 2.20462

    @staticmethod
    def lbs_to_kg(lbs: float) -> float:
        return lbs / 2.20462

    @staticmethod
    def km_to_miles(km: float) -> float:
        return km * 0.621371

    @staticmethod
    def miles_to_km(miles: float) -> float:
        return miles / 0.621371

File: Java/03-test-generation/Calculator.java
import java.util.*;

/**
 * Demo 3 — Test Generation (Refactored to Java)
 * ==============================================
 * This module has fully-implemented business logic but ZERO tests.
 * Use your testing framework (JUnit 5, TestNG) to generate a comprehensive test suite.
 *
 * Exercises:
 *   1. Create a test class → Generate unit tests for Calculator
 *   2. Test single methods with edge cases
 *   3. Add tests for UnitConverter static methods
 *   4. Run tests with: mvn test or gradle test
 */

/**
 * A calculator with memory, history, and scientific operations.
 */
public class Calculator {

    private double memory;
    private List<String> history;

    public Calculator() {
        this.memory = 0.0;
        this.history = new ArrayList<>();
    }

    public double add(double a, double b) {
        double result = a + b;
        record(a + " + " + b + " = " + result);
        return result;
    }

    public double subtract(double a, double b) {
        double result = a - b;
        record(a + " - " + b + " = " + result);
        return result;
    }

    public double multiply(double a, double b) {
        double result = a * b;
        record(a + " * " + b + " = " + result);
        return result;
    }

    public double divide(double a, double b) {
        if (b == 0) {
            throw new ArithmeticException("Cannot divide by zero");
        }
        double result = a / b;
        record(a + " / " + b + " = " + result);
        return result;
    }

    public double power(double base, double exponent) {
        double result = Math.pow(base, exponent);
        record(base + " ^ " + exponent + " = " + result);
        return result;
    }

    public double sqrt(double value) {
        if (value < 0) {
            throw new IllegalArgumentException("Cannot compute square root of a negative number");
        }
        double result = Math.sqrt(value);
        record("√" + value + " = " + result);
        return result;
    }

    /**
     * Return {@code percent}% of {@code value}. E.g. percentage(200, 15) → 30.0
     */
    public double percentage(double value, double percent) {
        double result = value * (percent / 100);
        record(percent + "% of " + value + " = " + result);
        return result;
    }

    /**
     * Save a value to memory.
     */
    public void store(double value) {
        this.memory = value;
    }

    /**
     * Retrieve the stored memory value.
     */
    public double recall() {
        return this.memory;
    }

    /**
     * Clear the memory.
     */
    public void clearMemory() {
        this.memory = 0.0;
    }

    /**
     * Get a copy of the calculation history.
     */
    public List<String> getHistory() {
        return new ArrayList<>(this.history);
    }

    /**
     * Clear the calculation history.
     */
    public void clearHistory() {
        this.history.clear();
    }

    /**
     * Record a calculation entry in history.
     */
    private void record(String entry) {
        this.history.add(entry);
    }
}

/**
 * Converts between common units.
 */
public class UnitConverter {

    private UnitConverter() {
        // Static utility class, prevent instantiation
    }

    public static double celsiusToFahrenheit(double c) {
        return (c * 9 / 5) + 32;
    }

    public static double fahrenheitToCelsius(double f) {
        return (f - 32) * 5 / 9;
    }

    public static double kgToLbs(double kg) {
        return kg * 2.20462;
    }

    public static double lbsToKg(double lbs) {
        return lbs / 2.20462;
    }

    public static double kmToMiles(double km) {
        return km * 0.621371;
    }

    public static double milesToKm(double miles) {
        return miles / 0.621371;
    }
}
